kube-runtime/src/manager.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{
+    controller::{controller, trigger_owners, trigger_self, trigger_with, Context, Error, ReconcilerAction},
+    reflector::{
+        reflector,
+        store::{Store, Writer},
+        ObjectRef,
+    },
+    utils::{try_flatten_applied, try_flatten_touched},
+    watcher::{self, watcher},
+};
+use futures::{stream, Stream, TryFuture};
+use kube::api::{Api, ListParams, Meta};
+use serde::de::DeserializeOwned;
+use std::pin::Pin;
+
+/// A builder for controller
+pub struct ControllerBuilder<K>
+where
+    K: Clone + Meta + 'static,
+{
+    // NB: Need to Unpin for stream::select_all
+    // TODO: get an arbitrary std::error::Error in here?
+    selector: Vec<Pin<Box<dyn Stream<Item = Result<ObjectRef<K>, watcher::Error>>>>>,
+    reader: Store<K>,
+}
+
+impl<K> ControllerBuilder<K>
+where
+    K: Clone + Meta + DeserializeOwned + 'static,
+{
+    /// Create a ControllerBuilder on a type `K`
+    ///
+    /// Configure `ListParams` and `Api` so you only get reconcile events
+    /// for the correct Api scope (cluster/all/namespaced), or ListParams subset
+    ///
+    /// A writer is exposed for convenience so you can peak into the main reflector's state.
+    pub fn new(owned_api: Api<K>, lp: ListParams, writer: Writer<K>) -> Self {
+        let reader = writer.as_reader();
+        let mut selector = vec![];
+        let self_watcher: Pin<Box<dyn Stream<Item = Result<ObjectRef<K>, watcher::Error>>>> = Box::pin(
+            trigger_self(try_flatten_applied(reflector(writer, watcher(owned_api, lp)))),
+        );
+        selector.push(self_watcher);
+        Self { selector, reader }
+    }
+
+    /// Indicate child objets K owns and be notified when they change
+    ///
+    /// This type `CHILD` must have OwnerReferences set to point back to `K`.
+    /// You can customize the parameters used by the underlying watcher if
+    /// only a subset of `CHILD` entries are required.
+    /// The `api` must have the correct scope (cluster/all namespaces, or namespaced)
+    pub fn owns<CHILD: Clone + Meta + DeserializeOwned + 'static>(
```

Comment:
AFAIK the convention is `PascalCase` for generics. I guess we should at least try to be consistent?

Reply:
oh, of course, sure!